service/borrowerService.js

```diff
@@ -1,120 +1,181 @@
-      );
-    return;
-  }
-
-  dao
-    .getBorrower(loan.cardNo)
-    // Check if borrower exists
-    .then((result) => {
-      if (result.length > 0) return dao.getBookCopies(loan);
-      else res.status(400).send("That card number is not in our database. ");
-    })
-    // Check if branch has a copy of the book, if so remove one from branch and checkout book.
-    .then((result) => {
-      if (result.length > 0 && result[0].noOfCopies > 0) {
-        loan.dateOut = moment(Date.now()).format("YYYY-MM-DD HH:mm:ss");
-        let dueDate = new Date();
-        dueDate = dueDate.setDate(dueDate.getDate() + 7);
-        loan.dueDate = moment(dueDate).format("YYYY-MM-DD HH:mm:ss");
-        return dao.checkOutBook(loan, result[0].noOfCopies - 1);
-      } else
-        res
-          .status(400)
-          .send("There are no copies of that book in our branch.  ");
-    })
-    // Return the checked out book to the user
-    .then((result) => {
-      res.status(201);
-      res.format({
-        "application/json": function () {
-          res.send(loan);
-        },
-        "application/xml": function () {
-          res.send(jsontoxml(loan));
-        },
-        "text/plain": function () {
-          res.send(loan.toString());
-        },
-      });
-    })
-    .catch((err) => {
-      res.status(500).send("Internal Server Error");
-      logger.error(err);
-      return;
+exports.readBorrower = (cardNo) => {
+  let result;
+  return new Promise((resolve, reject) => {
+    db.beginTransaction(async function (err) {
+      if (err) {
+        logger.error(err);
+        throw err;
+      }
+
+      try {
+        result = await dao.readBorrower(db, cardNo);
+      } catch (err) {
+        logger.error(err);
+        reject(err);
+      }
+
+      try {
+        loansResult = await dao.readLoansByBorrower(db, cardNo);
+      } catch (err) {
+        logger.err(err);
+        reject(err);
+      }
+
+      if (result.length > 0) {
+        result = result[0];
+        result.loans = [];
+
+        loansResult.forEach((val) => {
+          result.loans.push(val);
+        });
+      }
+
+      resolve(result);
+    });
+  });
+};
+
+exports.checkNumOfCopies = (loan) => {
+  let result;
+  return new Promise((resolve, reject) => {
+    db.beginTransaction(async function (err) {
+      if (err) {
+        logger.error(err);
+        throw err;
+      }
+
+      try {
+        result = await dao.readBookCopies(db, loan);
+      } catch (err) {
+        logger.error(err);
+        reject(err);
+      }
+      resolve(result);
+    });
+  });
+};
+
+exports.checkOutBook = (loan, numOfCopies) => {
+  return new Promise((resolve, reject) => {
+    db.beginTransaction(function (err) {
+      if (err) {
+        logger.error(err);
+        throw err;
+      }
+
+      loan.dateOut = moment(Date.now()).format("YYYY-MM-DD HH:mm:ss");
+      let dueDate = new Date();
+      dueDate = moment(dueDate).add(7, "days");
```

Comment:
Depending on how fast these operations happen, this could cause the due date to be slightly more than 7 days after the date out. Not something that necessarily needs fixing though since the difference will be so small.